src/ImageSharp/Processing/DefaultImageProcessorContext{TPixel}.cs

```diff
@@ -64,26 +67,25 @@ public IImageProcessingContext ApplyProcessor(IImageProcessor processor, Rectang
         {
             if (!this.mutate && this.destination is null)
             {
-                // This will only work if the first processor applied is the cloning one thus
-                // realistically for this optimization to work the resize must the first processor
-                // applied any only up processors will take the double data path.
-                using (IImageProcessor<TPixel> specificProcessor = processor.CreatePixelSpecificProcessor(this.source, rectangle))
+                // When cloning an image we can optimize the processing pipeline by avoiding an unnecessary
+                // interim clone if the first processor in the pipeline is a cloning processor.
+                if (processor is CloningImageProcessor cloningImageProcessor)
```

Comment:
Feels like we should not use a specific implementation here.

Reply:
I did this as I think we're going too heavy on the interfaces and I want to allow us flexibility to change implementation hence my idea to remove `ICloningImageProcessor<TPixel>`. By using the concrete instance and proxying the `CreatePixelSpecificProcessor` I think that better lets us do that.

This optimization behavior is specific to `DefaultImageProcessorContext` anyway.

Reply:
Introduced `ICloningImageProcessor` and appropriate methods. Much cleaner now.